src/cosalib/aliyun.py

```diff
@@ -82,11 +89,70 @@ def aliyun_run_ore_replicate(build, args):
             } for region, val in ore_data.items()
         ])
 
+    # short circuit the rest of this: if we haven't uploaded to all the specified
+    # regions, we shouldn't be continuing to try to make them public
+    if upload_failed_in_region is not None:
+        raise Exception(f"Upload failed in {upload_failed_in_region} region")
+
     build.meta['aliyun'] = aliyun_img_data
     build.meta_write()
 
-    if upload_failed_in_region is not None:
-        raise Exception(f"Upload failed in {upload_failed_in_region} region")
+    # delay marking images public to allow the cloud to register the newly
+    # created images
+    if args.public:
+        make_public(build, args)
+
+
+# a dedicated function for marking images public to take advantage of the
+# tenacity retry decorator.
+# image uploads can take quite some time to show up on the cloud, so we are
+# being generous with the retries.
+@retry(reraise=True, before=before_log(logger, log.INFO),
+       wait=wait_exponential(multiplier=1, min=4, max=10),
+       stop=stop_after_attempt(5))
+def make_public(build, args):
+    build.refresh_meta()
+    aliyun_img_data = build.meta.get('aliyun', [])
+
+    make_public_args = [
+        'ore',
+        '--log-level', args.log_level,
+        'aliyun', 'make-public'
+    ]
+
+    if args.config:
+        make_public_args.extend(['--config-file', args.config])
+
+    skipped_pairs = []
+    for entry in aliyun_img_data:
+        # id == image id, name == region id
+        image_region_pair = [f"{entry['id']}:{entry['name']}"]
```

Comment:
Just want to make sure I understand here: the `make-public` command supports specifying multiple region/image pairs but we're only specifying one at a time here?

Reply:
Yeah, the way the `make-public` command works now is that the operation will fail if one region fails out of the list. I wanted to be more resilient to those kinds of failures (as I've had unreliable experiences with their cloud) so I'm calling them one at a time.  (It also lets me accumulate failures into the skipped list to retry them later)

Reply:
I think the way we deal with this in other places is to just make the command idempotent (which I think `make-public` already is. Then you don't really need to keep up with failures. Just retry it with all regions.